src/main/java/appeng/core/sync/packets/PacketNEIBookmark.java

```diff
@@ -0,0 +1,118 @@
+import appeng.api.storage.IMEMonitor;
+import appeng.api.storage.data.IAEItemStack;
+import appeng.container.implementations.ContainerMEMonitorable;
+import appeng.container.implementations.ContainerWirelessTerm;
+import appeng.core.sync.AppEngPacket;
+import appeng.core.sync.network.INetworkInfo;
+import appeng.util.InventoryAdaptor;
+import appeng.util.Platform;
+import appeng.util.item.AEItemStack;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+
+public class PacketNEIBookmark extends AppEngPacket {
+
+    private ItemStack bookmarkItem;
+
+    // automatic
+    public PacketNEIBookmark(final ByteBuf stream) throws IOException {
+        final ByteArrayInputStream bytes = new ByteArrayInputStream(stream.array());
+        bytes.skip(stream.readerIndex());
+        final NBTTagCompound comp = CompressedStreamTools.readCompressed(bytes);
+        if (comp != null) {
+            this.bookmarkItem = ItemStack.loadItemStackFromNBT(comp);
+        }
+    }
+
+    // api
+    public PacketNEIBookmark(final NBTTagCompound bookmarkItemComp) throws IOException {
+        final ByteBuf data = Unpooled.buffer();
+
+        final ByteArrayOutputStream bytes = new ByteArrayOutputStream();
+        final DataOutputStream outputStream = new DataOutputStream(bytes);
+
+        data.writeInt(this.getPacketID());
+
+        CompressedStreamTools.writeCompressed(bookmarkItemComp, outputStream);
+        data.writeBytes(bytes.toByteArray());
+
+        this.configureWrite(data);
+    }
+
+    @Override
+    public void serverPacketData(final INetworkInfo manager, final AppEngPacket packet, final EntityPlayer player) {
+        if (isInventoryFull(player)) return;
+        final EntityPlayerMP pmp = (EntityPlayerMP) player;
+        final Container con = pmp.openContainer;
+
+        if (con instanceof ContainerMEMonitorable monitorable) {
+            final IGridNode node = monitorable.getNetworkNode();
+            if (node != null) {
+                final IGrid grid = node.getGrid();
+                if (grid == null) {
+                    return;
+                }
+
+                final IStorageGrid inv = grid.getCache(IStorageGrid.class);
+                final IMEMonitor<IAEItemStack> storage = inv.getItemInventory();
+                final IEnergyGrid energy = grid.getCache(IEnergyGrid.class);
+                final BaseActionSource actionSource = monitorable.getActionSource();
+
+                final AEItemStack request = AEItemStack.create(bookmarkItem);
+                final IAEItemStack out = Platform.poweredExtraction(energy, storage, request, actionSource);
+                if (out != null) {
+                    final InventoryAdaptor adp = InventoryAdaptor.getAdaptor(player, ForgeDirection.UNKNOWN);
+                    ItemStack outItem = out.getItemStack();
+                    adp.addItems(outItem);
+                }
+            }
+        }
+        if (con instanceof ContainerWirelessTerm wirelessTerm) {
```

Comment:
any reason for `if` over `else if`?

Reply:
`ContainerWirelessTerm wirelessTerm` is technically an instance of `ContainerMEMonitorable` which is why the `else if` was never triggered as it went into the first statement.
A second `if` enforces Java to also evaluate it and there it actually did the pull logic as it was able to find the `node`.

I suppose this is another good argument for the getter method in the review so the second `if` can be dropped and everything is more generic under `ContainerMEMonitorable`.